guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/auth/TokenRESTService.java

```diff
@@ -215,41 +195,151 @@ public APIAuthToken createToken(@FormParam("username") String username,
             }
 
         } // end Authorization header fallback
-        
+
         // Build credentials
         Credentials credentials = new Credentials();
         credentials.setUsername(username);
         credentials.setPassword(password);
         credentials.setRequest(request);
         credentials.setSession(request.getSession(true));
-        
-        UserContext userContext;
-        try {
 
-            // Update existing user context if session already exists
-            if (existingSession != null)
-                userContext = authProvider.updateUserContext(existingSession.getUserContext(), credentials);
+        return credentials;
+
+    }
 
-            /// Otherwise, generate a new user context
-            else {
+    /**
+     * Attempts authentication against all AuthenticationProviders, in order,
+     * using the provided credentials. The first authentication failure takes
+     * priority, but remaining AuthenticationProviders are attempted. If any
+     * AuthenticationProvider succeeds, the resulting AuthenticatedUser is
+     * returned, and no further AuthenticationProviders are tried.
+     *
+     * @param credentials
+     *     The credentials to use for authentication.
+     *
+     * @return
+     *     The AuthenticatedUser given by the highest-priority
+     *     AuthenticationProvider for which the given credentials are valid.
+     *
+     * @throws GuacamoleException
+     *     If the given credentials are not valid for any
+     *     AuthenticationProvider, or if an error occurs while authenticating
+     *     the user.
+     */
+    private AuthenticatedUser authenticateUser(Credentials credentials)
+        throws GuacamoleException {
 
-                userContext = authProvider.getUserContext(credentials);
+        GuacamoleCredentialsException authFailure = null;
 
-                // Log successful authentication
-                if (userContext != null && logger.isInfoEnabled())
-                    logger.info("User \"{}\" successfully authenticated from {}.",
-                            userContext.self().getIdentifier(), getLoggableAddress(request));
+        // Attempt authentication against each AuthenticationProvider
+        for (AuthenticationProvider authProvider : authProviders) {
 
+            // Attempt authentication
+            try {
+                AuthenticatedUser authenticatedUser = authProvider.authenticateUser(credentials);
+                if (authenticatedUser != null)
+                    return authenticatedUser;
             }
 
-            // Request standard username/password if no user context was produced
-            if (userContext == null)
-                throw new GuacamoleInvalidCredentialsException("Permission Denied.",
-                        CredentialsInfo.USERNAME_PASSWORD);
+            // First failure takes priority for now
+            catch (GuacamoleCredentialsException e) {
+                if (authFailure == null)
```

Comment:
Because we want to try the other auth providers before giving up. If one auth provider throws an exception indicating that it wants different credentials, that shouldn't stop the checking of other auth providers to see if someone else is fine with what we've got.

The other exceptions will bail out, however.